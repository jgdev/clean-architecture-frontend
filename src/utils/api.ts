import { reactive, unref } from "vue";
import { FetchAction, PaginatedParams, PaginatedResult } from "../types";
import { authSession } from "../api";

const apiUrl = import.meta.env.VITE_API_BASE_URL;

export const createFetchAction = <T, K = any>(
  apiEndpoint: string,
  defaultValue?: T,
  transform?: {
    response?: (body: any) => any;
  }
): FetchAction<T, K> => {
  const state = reactive({
    result: defaultValue,
    error: undefined,
  }) as FetchAction<T, K>;

  const fetchAction = (defaultOptions: RequestInit) => {
    return async (
      options?: RequestInit & {
        data?: K;
      }
    ): Promise<void> => {
      const requestOptions: RequestInit = {
        ...defaultOptions,
        ...options,
      };

      requestOptions.headers = {
        ...(requestOptions.headers || {}),
        ...(authSession.result && {
          "X-Session-Id": authSession.result,
        }),
      };

      state.loading = true;

      return fetch(apiUrl + apiEndpoint, {
        ...requestOptions,
        ...(requestOptions?.method !== "GET" &&
          options?.data && {
            body: JSON.stringify(options.data),
            headers: {
              ...(requestOptions.headers || {}),
              "Content-Type": "application/json",
            },
          }),
      })
        .then(async (res) => {
          const body = await res.json();
          if (!res.ok) return Promise.reject(body);
          return body as K;
        })
        .then(({ result }: any) => {
          state.result =
            (transform?.response && transform.response(result)) || result;
          return result;
        })
        .catch((err) => {
          state.error = err;
          return Promise.reject(err);
        })
        .finally(() => {
          state.loading = false;
        });
    };
  };

  state.delAction = fetchAction({ method: "DEL" });
  state.postAction = fetchAction({ method: "POST" });
  state.getAction = fetchAction({ method: "GET" });

  return unref(state);
};

export const createFetchResourcesAction = <T>(
  apiEndpoint: string,
  transform?: {
    response?: (body: any) => any;
  }
): FetchAction<PaginatedResult<T>> => {
  const state = createFetchAction<PaginatedResult<T>, PaginatedParams>(
    apiEndpoint,
    {
      result: [],
      limit: 0,
      searchParams: undefined,
      skip: 0,
    },
    transform
  );

  return state;
};
