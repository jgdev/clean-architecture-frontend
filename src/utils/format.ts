export const formatAmount = (amount: number) => {
  return amount.toLocaleString("en-US", {
    style: "currency",
    currency: "USD",
    currencyDisplay: "symbol",
  });
};

export const limitCharacters = (str: string, limit: number) => {
  return str.length > limit ? str.substring(0, limit) + '...' : str
}