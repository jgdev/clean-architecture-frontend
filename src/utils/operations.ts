export const stringGeneratorOptions = [
  ["Alphabetic", true],
  ["Uppercase", true],
  ["Lowercase", true],
  ["Numbers", true],
  ["Symbols", true],
];

export const mapConfiguration: { [key: string]: any } = {
  addition: {
    value: "addition",
    name: "Addition",
    input: {
      type: "number",
      multiple: true,
      minInput: 2,
      getLabel: (index: number) => (index === 0 ? "This number" : "Plus"),
      getPlaceholder: (index: number) => `Example: ${index + 1}`,
    },
  },
  subtraction: {
    value: "subtraction",
    name: "Subtract",
    input: {
      type: "number",
      multiple: true,
      minInput: 2,
      getLabel: (index: number) =>
        index === 0 ? "This number" : "Substracting",
      getPlaceholder: (index: number) => `Example: ${index + 1}`,
    },
  },
  multiplication: {
    value: "multiplication",
    name: "Multiply",
    input: {
      type: "number",
      multiple: true,
      minInput: 2,
      getLabel: (index: number) =>
        index === 0 ? "This number" : "Multiplied by",
      getPlaceholder: (index: number) => `Example: ${index + 1}`,
    },
  },
  division: {
    value: "division",
    name: "Divide",
    input: {
      type: "number",
      multiple: true,
      minInput: 2,
      getLabel: (index: number) => (index === 0 ? "This number" : "Divided by"),
      getPlaceholder: (index: number) => `Example: ${index + 1}`,
    },
  },
  square_root: {
    value: "square_root",
    name: "Square root",
    input: {
      type: "number",
      multiple: false,
      getLabel: () => "Square root of ...",
      getPlaceholder: () => "Type the base number",
    },
  },
  random_string: {
    value: "random_string",
    name: "Random String",
    input: {
      type: "number",
      multiple: false,
      options: stringGeneratorOptions,
      getLabel: () => "String length",
      getPlaceholder: () => "Example: 10",
    },
  },
  random_string_v2: {
    value: "random_string_v2",
    name: "Random String (Random.org)",
    input: {
      type: "number",
      multiple: false,
      options: stringGeneratorOptions,
      maxLength: 32,
      getLabel: () => "String length",
      getPlaceholder: () => "Example: 10",
    },
  },
};
