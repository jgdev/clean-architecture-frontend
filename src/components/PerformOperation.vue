<script lang="ts" setup>
import { ref, reactive, watch, computed } from 'vue'
import { Modal, Button, Spinner, Select } from 'flowbite-vue';
import FormOperationInput from './FormOperationInput.vue';
import { FetchResources, Operation } from '../types';
import { mapConfiguration } from '../utils/operations';

type Props = {
  onClose: () => void
  onSubmit: (operationId: string, args: any[]) => Promise<void>
  operations: FetchResources<Operation>
}

const props = defineProps<Props>()

const operationsConfig = computed(() => (props.operations.result?.result || []).map(operation => mapConfiguration[operation.type]).sort((a: Operation, b: Operation) => {
  if (a.name < b.name) return 1;
  if (b.name < a.name) return -1;
  return 0;
}))

const form = ref<HTMLFormElement>()
const loading = ref(false)
const operationSelected = ref('')
const error = ref('')
const state = reactive({
  inputs: {},
  options: {}
})

const onClose = () => {
  if (loading.value) return;
  props.onClose();
}

const onSubmit = (e: Event) => {
  e.preventDefault();
  const inputArgs = Object.values(state.inputs).map((v: any) => Number(v?.value))
  const optionsArgs = Object.values(state.options).map((v: any) => v.value);
  loading.value = true
  error.value = ''

  props.onSubmit(operationSelected.value, [...inputArgs, ...optionsArgs]).then(() => {
    loading.value = false;
    onClose()
  }).catch(err => {
    loading.value = false;
    error.value = err.detail || err.msg || err
  })
}

const handleSubmit = () => {
  form.value?.requestSubmit()
}

const handleInputsChange = (inputs: any, options: any) => {
  state.inputs = inputs;
  state.options = options;
  error.value = ''
}

watch(operationSelected, () => {
  error.value = ''
})
</script>

<template>
  <Modal @close="onClose" size="sm" :popup="true">
    <template #header>
      <h1>Perform operation</h1>
    </template>
    <template #body>
      <form @submit="onSubmit" ref="form">
        <Select size="sm" placeholder="Operation type" :disabled="loading" class="mb-4" :options="operationsConfig"
          v-model="operationSelected" required />
        <div class="flex flex-col gap-4">
          <FormOperationInput v-if="!!operationSelected"
            :get-input="() => operationsConfig.find(operation => operation.value === operationSelected)?.input || null"
            :update-input-values="(inputs: any, options: any) => handleInputsChange(inputs, options)" :loading="loading"
            :error="error" />
        </div>
      </form>
    </template>
    <template #footer>
      <div class="flex justify-between">
        <Button color="light" @click="onClose" :disabled="loading">Cancel</Button>
        <Button type="submit" color="default" outline :disabled="loading" @click="handleSubmit">
          <div class="flex justify-center items-center gap-2">
            {{ loading ? 'Performing ...' : 'Submit' }}
            <Spinner color="blue" size="4" v-if="loading" />
          </div>
        </Button>
      </div>
    </template>
  </Modal>
</template>