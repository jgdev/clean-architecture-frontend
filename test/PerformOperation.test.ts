import { mount } from "@vue/test-utils";
import PerformOperation from "../src/components/PerformOperation.vue";
import {
  FetchAction,
  FetchResources,
  User,
  Record,
  Operation,
} from "../src/types";
import { createFakeAction, createFakeResources } from "./utils";

describe("PerformOperation", () => {
  let user: FetchAction<User>;
  let authSession: FetchAction<string>;
  let authLogout: FetchAction<void>;
  let records: FetchResources<Record>;
  let operations: FetchResources<Operation>;
  let deleteRecord: FetchAction<void>;
  let props = {};
  let onClose = vi.fn(() => Promise.resolve());
  let onSubmit = vi.fn(() => Promise.resolve());

  beforeEach(() => {
    user = createFakeAction<User>({
      email: "test@test",
      balance: 200,
    });
    authSession = createFakeAction<string>();
    authLogout = createFakeAction<void>();
    records = createFakeResources<Record>();
    operations = createFakeResources<Operation>();
    props = {
      user,
      authSession,
      authLogout,
      records,
      operations,
      onClose,
      onSubmit,
    };
  });

  test("should render properly", () => {
    const wrapper = mount(PerformOperation, { props });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should disable controls while loading", async () => {
    const wrapper = mount(PerformOperation, { props });
    console.log(wrapper.find('button[type="submit"]').text());
    expect(wrapper.find("button").text()).toBe("Submit");
    await wrapper.find("form").trigger("submit");
    expect(wrapper.find('button[type="submit"]').text()).toBe("Performing ...");
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should display addition operation properly", () => {
    throw new Error("not implemented");
  });

  test("should submit addition operation properly", () => {
    throw new Error("not implemented");
  });

  test("should display subtraction operation properly", () => {
    throw new Error("not implemented");
  });

  test("should submit subtraction operation properly", () => {
    throw new Error("not implemented");
  });

  test("should display multiplication operation properly", () => {
    throw new Error("not implemented");
  });

  test("should submit multiplication operation properly", () => {
    throw new Error("not implemented");
  });

  test("should display division operation properly", () => {
    throw new Error("not implemented");
  });

  test("should submit division operation properly", () => {
    throw new Error("not implemented");
  });

  test("should display square_root operation properly", () => {
    throw new Error("not implemented");
  });

  test("should submit square_root operation properly", () => {
    throw new Error("not implemented");
  });

  test("should display string_generator operation properly", () => {
    throw new Error("not implemented");
  });

  test("should submit string_generator operation properly", () => {
    throw new Error("not implemented");
  });
});
