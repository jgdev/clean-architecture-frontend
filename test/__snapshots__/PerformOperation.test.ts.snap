// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PerformOperation > should render properly 1`] = `
"<div user=\\"[object Object]\\" authsession=\\"[object Object]\\" authlogout=\\"[object Object]\\" records=\\"[object Object]\\">
  <div class=\\"bg-gray-900 bg-opacity-50 dark:bg-opacity-80 fixed inset-0 z-40\\"></div>
  <div tabindex=\\"-1\\" class=\\"overflow-y-auto overflow-x-hidden fixed top-0 right-0 left-0 z-50 w-full md:inset-0 h-modal md:h-full justify-center items-center flex\\">
    <div class=\\"relative p-4 w-full h-full md:h-auto max-w-sm\\">
      <div class=\\"relative bg-white rounded-lg shadow dark:bg-gray-700\\">
        <div class=\\"p-4 rounded-t flex justify-between items-center border-b border-gray-200 dark:border-gray-600\\">
          <h1>Perform operation</h1><button type=\\"button\\" class=\\"text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm p-1.5 ml-auto inline-flex items-center dark:hover:bg-gray-600 dark:hover:text-white\\"><svg class=\\"w-5 h-5\\" fill=\\"currentColor\\" viewBox=\\"0 0 20 20\\" xmlns=\\"http://www.w3.org/2000/svg\\">
              <path fill-rule=\\"evenodd\\" d=\\"M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z\\" clip-rule=\\"evenodd\\"></path>
            </svg></button>
        </div>
        <div class=\\"p-6\\">
          <form>
            <div class=\\"mb-4\\" required=\\"\\">
              <!----><select class=\\"w-full text-sm bg-gray-50 border-[1px] border-gray-300 rounded-lg focus:ring-primary-500 focus:border-primary-500 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-primary-500 dark:focus:border-primary-500 p-2\\">
                <option disabled=\\"\\" value=\\"\\">Operation type</option>
              </select>
            </div>
            <div class=\\"flex flex-col gap-4\\">
              <!--v-if-->
            </div>
          </form>
        </div>
        <div class=\\"p-6 rounded-b border-gray-200 border-t dark:border-gray-600\\">
          <div class=\\"flex justify-between\\"><button class=\\"text-gray-900 bg-white border-gray-300 focus:outline-none focus:ring-4 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 hover:bg-gray-100 dark:hover:border-gray-600 px-4 py-2 text-gray-900 bg-white border border-gray-300 focus:outline-none focus:ring-4 focus:ring-gray-200 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 hover:bg-gray-100 dark:hover:border-gray-600 text-sm px-4 py-2\\">
              <!----><span class=\\"\\"><!---->Cancel<!----></span>
              <!---->
            </button><button class=\\"text-blue-700 border-blue-700 focus:ring-4 focus:outline-none font-medium rounded-lg dark:border-blue-500 dark:text-blue-500 dark:focus:ring-blue-800 hover:text-white hover:bg-blue-800 dark:hover:text-white dark:hover:bg-blue-600 px-4 py-2 text-blue-700 border border-blue-700 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm text-center dark:border-blue-500 dark:text-blue-500 dark:focus:ring-blue-800 hover:text-white hover:bg-blue-800 dark:hover:text-white dark:hover:bg-blue-600 text-sm px-4 py-2\\" type=\\"submit\\">
              <!----><span class=\\"\\"><!----><div class=\\"flex justify-center items-center gap-2\\">Submit <!--v-if--></div><!----></span>
              <!---->
            </button></div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
